src/volume_rendering/volume_render_layer.ts

```diff
@@ -220,7 +236,63 @@ export class VolumeRenderingRenderLayer extends PerspectiveViewRenderLayer {
           builder.addFragmentCode(`
 #define VOLUME_RENDERING true
 `);
+          let glsl_rgbaEmit = glsl_emitRGBAVolumeRendering;
+          let glsl_finalEmit = `
+  emitAccumAndRevealage(outputColor, 1.0 - revealage, 0u);
+`;
+          let glsl_emitIntensity = `
+void emitIntensity(float value) {
+}
+`;
 
+          let glsl_continualEmit = ``;
+          if (isProjectionMode(shaderParametersState.mode)) {
+            builder.addFragmentCode(`
+float userIntensity = -1.0;
+float savedDepth = 0.0;
+float savedIntensity = 0.0;
+`);
+            const emitFunctions = `
+void emitIntensity(float value) {
+  userIntensity = value;
+}
+float getIntensity() {
+  float userIntensitySet = step(-0.9, userIntensity);
+  return convertIntensity(mix(defaultMaxProjectionIntensity, userIntensity, userIntensitySet));
+}
+`;
+            glsl_emitIntensity = `
+float convertIntensity(float value) {
+  return clamp(value, 0.0, 1.0);
+}
+${emitFunctions}
+`;
+            if (shaderParametersState.mode === VolumeRenderingModes.MIN) {
+              glsl_emitIntensity = `
+float convertIntensity(float value) {
+  return clamp(1.0 - value, 0.0, 1.0);
+}
+${emitFunctions}
+`;
+            }
+            glsl_rgbaEmit = `
+void emitRGBA(vec4 rgba) {
+  float newIntensity = getIntensity();
+  float intensityChanged = step(savedIntensity, newIntensity - 0.00001);
+  float alpha = clamp(rgba.a, 0.0, 1.0);
+  outputColor = mix(outputColor, vec4(rgba.rgb * alpha, alpha), intensityChanged);
+  savedIntensity = mix(savedIntensity, newIntensity, intensityChanged); 
+  savedDepth = mix(savedDepth, depthAtRayPosition, intensityChanged);
+  userIntensity = -1.0;
```

Comment:
I think it might be clearer and simpler to check the intensity in the outer loop that calls userMain -- that way the user can call emitIntensity before or after calling emitRGBA